devtools/server/actors/replay/connection.js

```diff
@@ -1082,6 +1111,137 @@ Services.ppmm.addMessageListener("RecordingStarting", {
   },
 });
 
+function getChannelRecording(channel) {
+  for (const [frameLoader, entry] of recordings) {
+    if (frameLoader.browsingContext && frameLoader.browsingContext === channel.loadInfo.browsingContext) {
+      return entry.recording || null;
+    }
+  }
+
+  return null;
+}
+function ensureHttpChannel(channel) {
+  if (
+    !(channel instanceof Ci.nsIHttpChannel) ||
+    !(channel instanceof Ci.nsIClassifiedChannel)
+  ) {
+    return null;
+  }
+
+  channel = channel.QueryInterface(Ci.nsIHttpChannel);
+  channel = channel.QueryInterface(Ci.nsIClassifiedChannel);
+
+  if (channel instanceof Ci.nsIHttpChannelInternal) {
+    channel = channel.QueryInterface(Ci.nsIHttpChannelInternal);
```

Comment:
Just curious, but does this sequence of `QueryInterface` calls cause the channel object to hold the union of all of those interfaces, or is each `QueryInterface` taking the channel object down a type-specialization chain where each step has to be called on a channel of a particular abstract type?

Reply:
I 100% do not understand the underlying mechanism of `QueryInterface` in JS and also wish I knew the answer to this question. This is mostly just copy-pasted from other gecko code and me adding what I think I need.